lib/sharepoint/client.rb

```diff
@@ -11,6 +11,61 @@ module Sharepoint
   class Client
     FILENAME_INVALID_CHARS = '~"#%&*:<>?/\{|}'
 
+
+    class InvalidTokenError < StandardError
+    end
+
+    class Token
+      class << self
+        attr_accessor :token_expires_in
+        attr_accessor :token
+        attr_accessor :token_fetched_at
+      end
+    end
+
+    def token
+      return Token.token unless Token.token.nil? || token_expired?
+
+      Token.token = fetch_token
+    end
+
+    def auth
+      "Bearer #{token}"
+    end
+
+    def token_expired?
+      return true unless Token.token_fetched_at && Token.token_expires_in
+
+      Token.token_fetched_at + Token.token_expires_in < Time.now.utc.to_i
+    end
```

Comment:
is this a sort of JWT token to be used together with requests?

Reply:
yes

Reply:
Yes, but thanks to MS or IFAD we need a custom endpoint to retrieve the token